### ISSUE TITLE
ARPACK-NG is not deterministic

### ISSUE THREAD

**Issue by valtron** [scipy/scipy#10374]
### Reproducing code example:
```
import numpy as np
from scipy.sparse.linalg.eigen.arpack.arpack import _arpack

_arpack_solver = _arpack.__dict__['dsaupd']

for i in range(2):
	x1 = _arpack_solver(
		0, 'I', 'LM', 2, 0.0001,
		np.array([0, 0, 0, 1]),
		np.zeros((4, 4), np.float64),
		np.array([1, 0, 40, 1, 0, 0, 3, 0, 0, 0, 0], dtype = np.int32),
		np.zeros(11, np.int32), np.zeros((12,)), np.zeros((48,)), 1,
	)[2]
	x2 = _arpack_solver(
		1, 'I', 'LM', 2, 0.0001,
		np.array([0, 0, 0, 1]),
		np.zeros((4, 4), np.float64),
		np.array([1, 0, 40, 1, 0, 0, 3, 0, 0, 0, 0], dtype = np.int32),
		np.zeros(11, np.int32), np.zeros((12,)), np.zeros((48,)), 1,
	)[2]
	print(x1, x2)
```

The same line should be printed twice, but instead it prints:

```
[0. 0. 0. 1.] [ 3.95742464e-01  8.64960398e-04 -9.22720579e-01 -9.16567150e-01]
[0. 0. 0. 1.] [ 0.11759638 -0.29962625  0.90382696 -0.25045105]
```

### Scipy/Numpy/Python version information:
```
1.3.0 1.16.2 sys.version_info(major=3, minor=6, micro=7, releaselevel='final', serial=0)
```

### More Info
I also tried called the ARPACK function directly in C, and got the same result. Same problem on ARPACK-NG 3.4.0 and 3.5.0, but 3.6.0 gives correct (deterministic) output:

```
[0. 0. 0. 1.] [0. 0. 0. 0.]
[0. 0. 0. 1.] [0. 0. 0. 0.]
```

C code:
```
#include <stdio.h>

void dsaupd_(
	int *ido, char *bmat, int *n, char *which,
	int *nev, double *tol, double *resid, int *ncv,
	double *v, int *ldv, int *iparam, int *ipntr,
	double *workd, double *workl, int *lworkl, int *info
);

int main(int argc, char** argv) {
	for (int i = 0; i < 2; ++i) {
		{
			int    ido = 0;
			char*  bmat = "I";
			int    n = 4;
			char*  which = "LM";
			int    nev = 2;
			double tol = 0.0001;
			double resid[4] = { 0, 0, 0, 1 };
			int    ncv = 4;
			int    ldv = 4;
			double v[16] = { 0.0 };
			int    iparam[11] = { 1, 0, 40, 1, 0, 0, 3, 0, 0, 0, 0 };
			int    ipntr[11] = { 0 };
			double workd[12] = { 0 };
			double workl[48] = { 0 };
			int    lworkl = 48;
			int    info = 1;
			
			dsaupd_(
				&ido, bmat, &n, which, &nev, &tol, resid,
				&ncv, v, &ldv, iparam, ipntr, workd, workl, &lworkl, &info
			);
			
			printf("[%f %f %f %f]", resid[0], resid[1], resid[2], resid[3]);
		}
		
		{
			int    ido = 1;
			char*  bmat = "I";
			int    n = 4;
			char*  which = "LM";
			int    nev = 2;
			double tol = 0.0001;
			double resid[4] = { 0, 0, 0, 1 };
			int    ncv = 4;
			int    ldv = 4;
			double v[16] = { 0.0 };
			int    iparam[11] = { 1, 0, 40, 1, 0, 0, 3, 0, 0, 0, 0 };
			int    ipntr[11] = { 0 };
			double workd[12] = { 0 };
			double workl[48] = { 0 };
			int    lworkl = 48;
			int    info = 1;
			
			dsaupd_(
				&ido, bmat, &n, which, &nev, &tol, resid,
				&ncv, v, &ldv, iparam, ipntr, workd, workl, &lworkl, &info
			);
			
			printf(" [%f %f %f %f]\n", resid[0], resid[1], resid[2], resid[3]);
		}
	}
	
	return 0;
}
```

**Comment by pv** [comment]
Right, arpack 3.6.0 has some fixes for the initial vector generation whereas scipy is at 3.3.0. Unfortunately, some tests that worked with arpack<3.5.0 seem to fail with 3.6.0, which needs to be resolved before upgrading... #10375

Note that the results are generally in principle always nondeterministic, because if you don't explicitly specify the initial vector, ARPACK will generate a random initial vector.

**Comment by pv** [comment]
Also: why do you say the same line should be printed twice? You are not calling the function with the same parameters.

**Comment by valtron** [comment]
This example specifies the initial vector (`resid`, `info = 1`) so it should be deterministic.

There are two calls to `dsaupd_` because it takes two iterations for the non-deterministic result to show up. Not that the two calls are in a loop; which is why both lines in the output (each produced from two iterations of `dsaupd_`) should be the same.

**Comment by pv** [comment]
Right --- however, ARPACK still appears to intentionally generate and use random vectors in certain cases, not just for the initial vector. This is not something that has been addressed in ARPACK 3.7.0, see https://github.com/opencollab/arpack-ng/blob/master/SRC/dsaitr.f#L409

The difference in behavior between the 3.5.0 and 3.6.0 versions is then incidental --- the changes made are unrelated to this issue. I guess if you want the random seed behavior in arpack to be controllable, this should be raised with arpack-ng people.

**Comment by valtron** [comment]
Thanks, I was stuck figuring out where it was coming from!

**Comment by chiggum** [comment]
Just wondering if this is fixed.

**Comment by mckib2** [comment]
@mdhaber Is this the same behavior we saw [here](url)?

**Comment by ilayn** [comment]
This has been addressed by the bump in the ARPACK version to 3.9.1. Now the main is giving 

```
[0. 0. 0. 1.] [0. 0. 0. 0.]
[0. 0. 0. 1.] [0. 0. 0. 0.]
```

Hence closing this one.
